sync/l1verifier.go

```diff
@@ -0,0 +1,109 @@
+package sync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/clients"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/sourcegraph/conc/stream"
+)
+
+type ErrL1VerificationFailed struct {
+	Height uint64
+	Err    error
+}
+
+func (e ErrL1VerificationFailed) Error() string {
+	return fmt.Sprintf("L1 verification failed on block #%d with %s", e.Height, e.Err.Error())
+}
+
+// L1Verifier is a background worker that continously tries to detect the highest height that was verified
+// on L1.
+type L1Verifier struct {
+	Blockchain *blockchain.Blockchain
+	Gateway    *clients.GatewayClient
+	Wait       time.Time
+	log        utils.SimpleLogger
+}
+
+// NewL1Verifier creates a new L1Verifier
+func NewL1Verifier(bc *blockchain.Blockchain, gateway *clients.GatewayClient, log utils.SimpleLogger) *L1Verifier {
+	return &L1Verifier{
+		Blockchain: bc,
+		Gateway:    gateway,
+		log:        log,
+	}
+}
+
+// Run starts the L1Verifier.
+func (v *L1Verifier) Run(ctx context.Context) error {
+	errChan := make(chan ErrL1VerificationFailed)
+	l1Stream := stream.New().WithMaxGoroutines(runtime.NumCPU())
```

Comment:
We are not supposed to L1 verify blocks in parallel like we do in sync.

Assume our L2 height is 100 and L1 height is `nil`. The actual L1 verified height is 75, but we havent figured that out yet.

Try using a binary search like approach to find the latest L1 verified block. For example;
- GetBlock(100), and notice it is not L1 verified
- GetBlock(50), and notice it IS L1 verified, make `blockchain` know that at height 50 a block with hash of `H1` was verified
- GetBlock(75)  and notice it IS L1 verified, make `blockchain` know that at height 75 a block with hash of `H2` was verified
- GetBlock(87), not verified
- GetBlock(82), not verified

At some point this search will yield and L1Verifier will realize that 75 is the latest block that is L1 verified and at that point it can start to L1 verify blocks one by one. You dont need a `stream` here, just make sure `Run()` does not block the execution and can be cancelled with ctx.

Reply:
Done